block/submit.go

```diff
@@ -2,200 +2,210 @@ package block
+				pending = pendingBytes.Add(^(nConsumed - 1)) // subtract
 			}
+			counter.Nudge()
 		}
-		select {
-		case <-ctx.Done():
-			return
-		case size := <-m.producedSizeC:
-			total += size
-		}
-	}
-}
-
-// HandleSubmissionTrigger processes the sublayer submission trigger event. It checks if there are new blocks produced since the last submission.
-// If there are, it attempts to submit a batch of blocks. It then attempts to produce an empty block to ensure IBC messages
-// pass through during the batch submission process due to proofs requires for ibc messages only exist on the next block.
-// Finally, it submits the next batch of blocks and updates the sync target to the height of the last block in the submitted batch.
-func (m *Manager) HandleSubmissionTrigger() error {
-	// Load current sync target and height to determine if new blocks are available for submission.
+	})
 
-	startHeight := m.NextHeightToSubmit()
-	endHeightInclusive := m.State.Height()
+	return eg.Wait()
+}
 
-	if endHeightInclusive < startHeight {
-		return nil // No new blocks have been produced
+func (m *Manager) CreateAndSubmitBatchGetSizeEstimate(maxSize uint64) (uint64, error) {
+	b, err := m.CreateAndSubmitBatch(maxSize)
+	if b == nil {
+		return 0, err
 	}
+	return uint64(b.SizeBytesEstimate()), err
+}
 
-	nextBatch, err := m.CreateNextBatchToSubmit(startHeight, endHeightInclusive)
+// CreateAndSubmitBatch creates and submits a batch to the DA and SL.
+func (m *Manager) CreateAndSubmitBatch(maxSize uint64) (*types.Batch, error) {
+	b, err := CreateBatch(m.Store, maxSize, m.NextHeightToSubmit(), m.State.Height())
 	if err != nil {
-		return fmt.Errorf("create next batch to submit: %w", err)
+		return nil, fmt.Errorf("create batch: %w", err)
 	}
 
-	resultSubmitToDA := m.DAClient.SubmitBatch(nextBatch)
-	m.logger.Info("Submitted batch to DA", "start height", nextBatch.StartHeight, "end height", nextBatch.EndHeight)
-	if resultSubmitToDA.Code != da.StatusSuccess {
-		return fmt.Errorf("submit next batch to da: %s", resultSubmitToDA.Message)
+	if 0 == b.NumBlocks() {
+		return nil, nil
 	}
 
-	actualEndHeight := nextBatch.EndHeight
-
-	err = m.SLClient.SubmitBatch(nextBatch, m.DAClient.GetClientType(), &resultSubmitToDA)
-	if err != nil {
-		return fmt.Errorf("sl client submit batch: start height: %d: inclusive end height: %d: %w", startHeight, actualEndHeight, err)
+	if err := m.SubmitBatch(b); err != nil {
+		return nil, fmt.Errorf("submit batch: %w", err)
 	}
-	m.logger.Info("Submitted batch to SL.", "start height", nextBatch.StartHeight, "end height", actualEndHeight)
-
-	types.RollappHubHeightGauge.Set(float64(actualEndHeight))
-	m.LastSubmittedHeight.Store(actualEndHeight)
-	return nil
+	return b, nil
 }
 
-func (m *Manager) CreateNextBatchToSubmit(startHeight uint64, endHeightInclusive uint64) (*types.Batch, error) {
+func CreateBatch(store store.Store, maxBatchSize uint64, startHeight uint64, endHeightInclusive uint64) (*types.Batch, error) {
```

Comment:
why it;'s not `func (m *Manager)` method?

Reply:
done